File: origin/YuGiOhCard/Controller/SettingViewController.m
#import "SettingViewController.h"
#import "StringConsts.h"
#import "ConfigUtils.h"
#import "RootViewController.h"
#import "FileUtils.h"

@interface UIBackgroundImageCell()

@end

@implementation UIBackgroundImageCell

@end


@interface SettingViewController () {
    NSMutableArray * _backgrounds;
    CGSize _grid_size;
    NSString * _current_background;
    NSString * _document;
    NSFileManager * fmgr;
}

@end

@implementation SettingViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = RIGHT_MENU_SETTING;
    _grid_size = [self generateImageSize];
    _current_background = [ConfigUtils loadBackgroundImage];
    if (_current_background == nil || [_current_background isEqualToString:@""]) {
        _current_background = @"bg1";
    }
    
    _backgrounds = [[NSMutableArray alloc] init];
    for (int i=1; i<=9; i++) {
        [_backgrounds addObject:[NSString stringWithFormat:@"bg%d", i]];
    }
    _document = [FileUtils getDocumentPath];
    fmgr = [NSFileManager defaultManager];
    NSString * sizeStr = [NSString stringWithFormat:@"%.2f M", [FileUtils folderSizeAtPath:_document]];
    [self.btnSpace setTitle:sizeStr forState:UIControlStateNormal];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

#pragma mark - action
-(IBAction)cleanClicked:(id)sender {
    UIAlertView * alert = [[UIAlertView alloc] initWithTitle:STR_CONFIRM message:STR_CONFIRM_CLEAR_DOCUMENT delegate:self cancelButtonTitle:COMMON_CANCEL otherButtonTitles:COMMON_OK, nil];
    [alert show];
}

-(void) alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex {
    NSLog(@"button clicked: %ld", (long)buttonIndex);
    switch (buttonIndex) {
        case 1:
            [self removeDocumentFiles];
            [self.btnSpace setTitle:[NSString stringWithFormat:@"%.2f M", [FileUtils folderSizeAtPath:_document]] forState:UIControlStateNormal];
            break;
        default:
            break;
    }
}

#pragma mark - file

-(void) removeDocumentFiles {
    NSString * imgPath = [_document stringByAppendingPathComponent:@"image"];
    [fmgr removeItemAtPath:imgPath error:nil];
    if (![fmgr fileExistsAtPath:imgPath]) {
        [fmgr createDirectoryAtPath:imgPath withIntermediateDirectories:YES attributes:nil error:nil];
    }
}

#pragma mark - collection view delegate

-(NSInteger) numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return 1;
}

-(NSInteger) collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return _backgrounds.count;
}

-(UIEdgeInsets) collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout insetForSectionAtIndex:(NSInteger)section {
    return UIEdgeInsetsMake(0, 0, 0, 0);
}

-(UICollectionViewCell *) collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    UIBackgroundImageCell * cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"Cell" forIndexPath:indexPath];
    cell.imgName = _backgrounds[indexPath.row];
    [cell.img setImage:[UIImage imageNamed:_backgrounds[indexPath.row]]];
    [cell.selectMark setHidden:![_current_background isEqualToString:cell.imgName]];
    return cell;
}

-(BOOL) collectionView:(UICollectionView *)collectionView shouldSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    return YES;
}

-(void) collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    [collectionView deselectItemAtIndexPath:indexPath animated:YES];
    _current_background = _backgrounds[indexPath.row];
    [ConfigUtils saveBackgroundImage:_current_background];
    RootViewController * root = [RootViewController getInstance];
    if (root) {
        root.backgroundImage = [UIImage imageNamed:_current_background];
    }
    [collectionView reloadData];
}

-(CGSize) collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    return _grid_size;
}

-(CGSize) generateImageSize {
    return CGSizeMake(89, 166);
}

@end

File: swift/YuGiOhCard/Controller/SettingViewController.swift
import UIKit

class UIBackgroundImageCell : UICollectionViewCell {
    var imgName: NSString?
    @IBOutlet var img: UIImageView?
    @IBOutlet var selectMark: UIImageView?
}


class SettingViewController: UIViewController, UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout, UIAlertViewDelegate {

    @IBOutlet var btnSpace: UIButton?
    @IBOutlet var colImg: UICollectionView?
    @IBOutlet var btnSource: UIButton?
    
    var _backgrounds: NSMutableArray?
    var _grid_size: CGSize?
    var _current_background: String?
    var _document: String?
    var fmgr: FileManager?
    
    var inited = false
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        if (!inited) {
            inited = true
            for temp in self.view.subviews {
                if (temp is UILabel) || (temp is UIButton) || (temp is UICollectionView) {
                    UIUtils.scaleComponent(view: temp)
                }
            }
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        UIUtils.setStatusBar(light: true)
        UIUtils.setNavBar(nav: self.navigationController!.navigationBar)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        inited = false
        self.title = RIGHT_MENU_SETTING
        _grid_size = self.generateImageSize()
        _current_background = ConfigUtils.loadBackgroundImage()
        if (_current_background == nil || _current_background! == "") {
            _current_background = "bg1"
        }
        
        _backgrounds = NSMutableArray()
        for i in 1 ... 9 {
            _backgrounds!.add("bg\(i)")
        }
        _document = FileUtils.getDocumentPath()
        fmgr = FileManager.default()
        let sizeStr = NSString(format: "%.2f M", FileUtils.folderSizeAtPath(folderPath: _document!))
        self.btnSpace!.setTitle(sizeStr as String, for: [])
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    @IBAction func cleanClicked(sender: AnyObject) {
        let alert = UIAlertView(title: STR_CONFIRM, message: STR_CONFIRM_CLEAR_DOCUMENT, delegate: self, cancelButtonTitle: COMMON_CANCEL, otherButtonTitles: COMMON_OK)
        alert.show()
    }
    
    func alertView(_ alertView: UIAlertView, clickedButtonAt buttonIndex: Int) {
        switch (buttonIndex) {
        case 1:
            self.removeDocumentFiles()
            self.btnSpace!.setTitle(NSString(format: "%.2f M", FileUtils.folderSizeAtPath(folderPath: _document!)) as String,for:[])
        default:
            break
        }
    }
    
    func removeDocumentFiles() {
        let imgPath = "\(_document!)/image"
        do {
            try fmgr!.removeItem(atPath: imgPath)
        } catch _ {
        }
        if (!fmgr!.fileExists(atPath: imgPath)) {
            do {
                try fmgr!.createDirectory(atPath: imgPath, withIntermediateDirectories: true, attributes: nil)
            } catch _ {
            }
        }
    }
    
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 1
    }

    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return _backgrounds!.count
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, insetForSectionAt section: Int) -> UIEdgeInsets {
        return UIEdgeInsetsMake(0, 0, 0, 0)
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "Cell", for:indexPath as IndexPath) as! UIBackgroundImageCell
        cell.imgName = _backgrounds![indexPath.row] as! String
        cell.img!.image = UIImage(named: cell.imgName as! String)
        cell.selectMark!.isHidden = _current_background! != cell.imgName
        return cell
    }
    
    func collectionView(_ collectionView: UICollectionView, shouldSelectItemAt indexPath: IndexPath) -> Bool {
        return true
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        collectionView.deselectItem(at: indexPath as IndexPath, animated:true)
        _current_background = _backgrounds![indexPath.row] as? String
        ConfigUtils.saveBackgroundImage(imgName: _current_background!)
        let nc = NotificationCenter.default()
        nc.post(name: "Notification_ChangeBackground" as NSNotification.Name, object: _current_background!)
        collectionView.reloadData()
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        return _grid_size!
    }
    
    func generateImageSize() -> CGSize {
        let (screenWidth, _) = UIUtils.getScreenSize()
        var scale: CGFloat = 1.0
        if (screenWidth > 320) {
            scale = screenWidth / 320
        }
        let w = (screenWidth - ((13 * 4) * scale)) / 3
        let h = 166 * w / 89
        return CGSize(width: w, height: h)
    }

    @IBAction func sourceClicked(sender: AnyObject) {
        UIApplication.shared().openURL(NSURL(string: URL_GITHUB)! as URL)
    }

}
